modules/raft/src/main/java/org/apache/ignite/raft/jraft/rpc/impl/ActionRequestProcessor.java

```diff
@@ -71,13 +75,23 @@ public void handleRequest(RpcContext rpcCtx, ActionRequest request) {
 
         JraftServerImpl.DelegatingStateMachine fsm = (JraftServerImpl.DelegatingStateMachine) node.getOptions().getFsm();
 
-        // Apply a filter before committing to STM.
-        fsm.getListener().onBeforeApply(request.command());
+        if (fsm.getListener().atomicOnBeforeApplyAndWrite(request.command()) && request.command() instanceof WriteCommand) {
+            // Synchronizing per group.
+            synchronized (internizer.intern(request.groupId())) {
```

Comment:
I really think that this "internizer" class is an overkill. Can we synchronize on the "node" object instead?

Reply:
Yes, we could, but someone else could do the same in another part of the code which might lead to less parallelizm (unnecessarily). Another solution is to add a monitor object to the `NodeImpl` (one per instance), but we would have to add an accessor for it to the `Node` interface, and this also seems non-desirable.

Reply:
> but someone else could do the same in another part of the code which might lead to less parallelizm

I don't think it's an issue

Another solution would be to add an explicit string -> object map to hold per-group locks (which is what this internizer does)

Reply:
I moved the map from internizer to the processor and removed internizer implementation